lib/model/message.dart

```diff
@@ -0,0 +1,193 @@
+    for (int i = 0; i < messages.length; i++) {
+      final message = messages[i];
+      messages[i] = this.messages.putIfAbsent(message.id, () => message);
+    }
+  }
+
+  void handleMessageEvent(MessageEvent event) {
+    // If the message is one we already know about (from a fetch),
+    // clobber it with the one from the event system.
+    // See [fetchedMessages] for reasoning.
+    messages[event.message.id] = event.message;
+
+    for (final view in _messageListViews) {
+      view.handleMessageEvent(event);
+    }
+  }
+
+  void handleUpdateMessageEvent(UpdateMessageEvent event) {
+    _handleUpdateMessageEventContent(event);
+    // TODO(#150): Handle message moves.  The views' recipient headers
+    //   may need updating, and consequently showSender too.
+  }
+
+  void _handleUpdateMessageEventContent(UpdateMessageEvent event) {
+    final message = messages[event.messageId];
+    if (message == null) return;
+
+    // TODO(server-5): Cut this fallback; rely on renderingOnly from FL 114
+    final isRenderingOnly = event.renderingOnly ?? (event.userId == null);
+    if (event.editTimestamp != null && !isRenderingOnly) {
+      // A rendering-only update gets omitted from the message edit history,
+      // and [Message.lastEditTimestamp] is the last timestamp of that history.
+      // So on a rendering-only update, the timestamp doesn't get updated.
+      message.lastEditTimestamp = event.editTimestamp;
+    }
+
+    message.flags = event.flags;
+
+    if (event.renderedContent != null) {
+      assert(message.contentType == 'text/html',
+        "Message contentType was ${message.contentType}; expected text/html.");
+      message.content = event.renderedContent!;
+    }
+
+    if (event.isMeMessage != null) {
+      message.isMeMessage = event.isMeMessage!;
+    }
+
+    for (final view in _messageListViews) {
+      view.messageContentChanged(event.messageId);
+    }
+  }
+
+  void handleDeleteMessageEvent(DeleteMessageEvent event) {
+    // TODO handle DeleteMessageEvent, particularly in MessageListView
+  }
+
+  void handleUpdateMessageFlagsEvent(UpdateMessageFlagsEvent event) {
+    final isAdd = switch (event) {
+      UpdateMessageFlagsAddEvent()    => true,
+      UpdateMessageFlagsRemoveEvent() => false,
+    };
+
+    if (isAdd && (event as UpdateMessageFlagsAddEvent).all) {
+      for (final message in messages.values) {
+        message.flags.add(event.flag);
+      }
+
+      for (final view in _messageListViews) {
+        if (view.messages.isEmpty) continue;
+        view.notifyListenersUnconditionally();
+      }
+    } else {
+      for (final messageId in event.messages) {
+        final message = messages[messageId];
+        if (message == null) continue; // a message we don't know about yet
+
+        isAdd
+          ? message.flags.add(event.flag)
+          : message.flags.remove(event.flag);
+
+        for (final view in _messageListViews) {
+          view.notifyListenersIfMessagePresent(messageId);
```

Comment:
This can cause a lot of `notifyListeners` calls, if a bunch of messages get marked read or unread in a narrow the user is looking at.

Generally I think it's OK to have a little duplication of those calls — for example if some event can cause it to be called two or three times, and it'd take a lot of refactoring or bookkeeping to prevent that and call just once, then that's OK. If some listener callback is expensive enough to make that a problem, then it should have appropriate memoization, or defer the real work until the next frame starts, or something, in order to mitigate that.

But this is potentially hundreds of times, so that gets me a bit concerned. Instead we could pass the whole list to each MessageListView as a batch, and let it decide whether to notify a single time.

Reply:
Good point, yeah.